### ISSUE TITLE
add support for additional static routes to each interface

### ISSUE THREAD

**Issue by kaniini** [ifupdown-ng/ifupdown-ng#42]
@kunkku has an addon for traditional ifupdown (which is supported by ifupdown-ng) that allows for additional static routes to be added to each device.  It looks like this:

```
iface eth0
   route 1.2.3.4/24,5.6.7.8/24
   rule ???
```

I'm not exactly how the `rule` part works.

Using ifupdown-ng executors, we can improve the usefulness of this feature by using multiple properties and ifquery (as we do with `address` and `gateway` properties).

So, this issue is about bikeshedding how that should look.

**Comment by BarbarossaTM** [comment]
How 'bout

```
iface eth0
    route 192.0.2.0/24   # device route
    route 10.0.0.0/8 172.16.0.1  # route with next-hop
```

I would prefer spaces instead of `,`

Is `rule` ment to be for PBR?

**Comment by kaniini** [comment]
Yes, rule is meant to be for PBR.

**Comment by kunkku** [comment]
The current syntax is:
```
iface eth0 inet static
    address 192.0.2.12
    netmask 255.255.255.0
    route 172.16.0.0/24,10.0.0.0/8 via 172.16.0.1
```

This translates to:
```
ip route add|del 172.16.0.0/24 src 192.0.2.12 dev eth0
ip route add|del 10.0.0.0/8 via 172.16.0.1 src 192.0.2.12 dev eth0
```
It is possible to pass arbitrary options supported by `ip route`. An example of a policy to route outgoing e-mail via `eth0` and the given gateway while limiting the path MTU:
```
iface eth0 inet static
    address 192.0.2.34
    netmask 255.255.255.0
    route default via 192.0.2.1 mtu 1200 table 123
    rule dport 25 table 123,dport 587 table 123
```
This translates to:
```
ip route add|del default via 192.0.2.1 mtu 1200 table 123 src 192.0.2.34 dev eth0
ip rule add|del dport 25 table 123
ip rule add|del dport 587 table 123
```
I agree specifying each route and rule on its own line would be clearer than the comma-separated syntax.

**Comment by BarbarossaTM** [comment]
Cool, thanks for the explanation! I agree that "one route/rule one line" would be cleaner.

Do you intend to set the `src` explicitly or did you just copy the output from `ip r` which had it set by the kernel?

**Comment by kunkku** [comment]
The current `ifupdown` addon appends `src` automatically according to the `address` directive. However, as `ifupdown-ng` supports defining multiple addresses, the behavior should be reconsidered.

**Comment by BarbarossaTM** [comment]
Indeed. I guess the best and yet simplest option would be to not set the `src` manually and let the user append `src 192.0.2.42` if setting a src IP is desired.

**Comment by kaniini** [comment]
I agree, that is probably what we should do.  Or perhaps have a property like `route-source`.

**Comment by BarbarossaTM** [comment]
> I agree, that is probably what we should do. Or perhaps have a property like `route-source`.

I would vote against that as you could only specify one source address for all routes. If someone wants to specify a src, just adding `src 192.0.2.42` to the route would do the same thing and allow different src addresses for different routes (if someone would want to do that, but I lately learned people do strange things).

**Comment by kunkku** [comment]
I believe that setting the source address to one of the interface's addresses is desired more often than not. Otherwise, the source address is usually set to the one facing the default gateway, resulting in asymmetric routing. And since interfaces typically have only one address per family (IPv4 or IPv6) configured, forcing the user to repeat the sole address on each `route` line would be annoying.

I suggest defaulting `src` to the first address of the appropriate family unless defined explicitly. This would be compatible with the current implementation.

**Comment by BarbarossaTM** [comment]
Please let's not add code for what the kernel does already.

Having multiple addresses even from different subnets is possible and is used in the wild for various setups and reasons. If we do this, we should do it generically and not tell people that they are doing it wrong and we don't support their setup.

About annoyance of repeating the src per route: I would assume people automating their config and if they need to set the src explicitly it implies that there are multiple IPs or a totally unrelated IP should be used. So that shouldn't happen that often. At least not in a statically configured setup.

**Comment by kunkku** [comment]
The kernel sets `src` for the routes created automatically when adding a new interface address, but not for the additional routes. If you want the source address to match `dev`, you must specify `src` unless there is another matching route with a shorter prefix which happens to have the correct `src`.

**Comment by kaniini** [comment]
I think it is desirable to automatically set `src`, for backwards compatibility with Alpine users setups.

We could add a config option to disable that for people who want to specify it explicitly.

**Comment by kaniini** [comment]
Kicking this to 0.10 since we don't fully yet have consensus.

**Comment by kaniini** [comment]
One last thing to consider: should this support be part of the `static` executor, or should we add an additional executor for this feature?

Would like to ship 0.10 this week so we make it into Alpine 3.13 -- so going to move this to 0.11 milestone regardless.

**Comment by BarbarossaTM** [comment]
I'd be fine with it being part of the static executor. If we call it `static-route` (which it is anyway), id would be trivial to do so, too ;-)

**Comment by kaniini** [comment]
I went with a separate executor (and thus vocabulary) and added compat glue to rewrite the route/rule statements to fit the new vocabulary.

This looks pretty good IMO.

```
nanabozho:~/ifupdown-ng$ cat tests/fixtures/static-routing.interfaces 
iface xlat-route
        address 192.0.2.12/24
        route 172.16.0.0/16,10.0.0.0/8 via 172.16.0.1

iface xlat-rule
        address 192.0.2.12/24
        route default via 192.0.2.1 mtu 1200 table 123
        rule dport 25 table 123,dport 587 table 123
nanabozho:~/ifupdown-ng$ ./ifquery -L -P -i tests/fixtures/static-routing.interfaces 
auto lo
iface lo
  use link
  use loopback

iface xlat-route
  use link
  use static
  address 192.0.2.12/24
  use routing
  routing-route 172.16.0.0/16
  routing-route 10.0.0.0/8 via 172.16.0.1

iface xlat-rule
  use link
  use static
  address 192.0.2.12/24
  use routing
  routing-route default via 192.0.2.1 mtu 1200 table 123
  routing-rule dport 25 table 123
  routing-rule dport 587 table 123
```

**Comment by kaniini** [comment]
there is still the matter of the `src` attribute.  on that topic, i don't yet have a good answer.

**Comment by BarbarossaTM** [comment]
Hm, I vote against `routing-*`. Reasons:
 *  `static-route` is more clear on what this is and uses the common name
 * For MPLS I will add `mpls-route` support which I intend to implement within the MPLS executor as we don't have to do magic do distinguish between IP routes and MPLS routes

**Comment by kaniini** [comment]
I'm open to using a different namespace, but want it separate from the `static` executor.  `static-route` won't work due to the namespacing rules.

**Comment by BarbarossaTM** [comment]
Hmmm, I see. From a namespace perspective this might make sense:

- `route-static`
- `route-rule`
- `route-mpls`

Unsure about the last bit though, it might still make sense to have it as `mpls-route`. What do you think?

**Comment by kaniini** [comment]
That seems agreeable to me.

**Comment by CosmicToast** [comment]
I was missing this and threw something up while waiting for the consensus and resolution.
Throwing it in here in case someone else might need it (while waiting for the official solution, of course).
This isn't compatible with the old ifupdown system, nor with the propositions here (I'm not aware of how to properly handle multiples of the same statements, so I just went for the comma-separated route).
<details>
<summary>/usr/libexec/ifupdown-ng/routes</summary>

```sh
#!/bin/sh
[ -z "${VERBOSE}" ] || set -x

# routes-static 1.2.3.0/24,10.0.0.0/8 via 1.2.3.4
# routes-rule dport 25 table 123,dport 587 table 123

# adds $3 to $1 if $1 does not contain $2
addif() {
	if [ "$1" = "${1%$2*}" ]; then
		echo $1 $3
	else
		echo $1
	fi
}
decidr() {
	echo "${1%/*}"
}
addsrc() {
	addif "$1" src "src $(decidr $IF_ADDRESS)"
}
adddev() {
	addif "$1" dev "dev $IFACE"
}

# $1: input string
# $2: delimeter
# $3: function to call
foreach() {
	list="$1"

	while [ -n "$list" ]; do
		line="${list%%$2*}"
		list="${list#*$2}"
		"$3" "$line"

		[ "$line" = "$list" ] && break
	done
}

# add a route
prep() {
	route="$1"
	route=$(addsrc "$route")
	route=$(adddev "$route")
	echo $route
}
routeadd() {
	${MOCK} ip route add $(prep "$1")
}
routedel() {
	${MOCK} ip route del $(prep "$1")
}

ruleadd() {
	${MOCK} ip rule add $1
}
ruledel() {
	${MOCK} ip rule del $1
}

case "$PHASE" in
up)
	foreach "$IF_ROUTES_STATIC" , routeadd
	foreach "$IF_ROUTES_RULE" , ruleadd
	;;
down)
	foreach "$IF_ROUTES_STATIC" , routedel
	foreach "$IF_ROUTES_RULE" , ruledel
	;;
*) exit 0 ;;
esac
```

</details>

This functions well enough for my (temporary) use-case, though users are obviously encouraged to modify it to their needs.
Note that while #116 remains unreleased, this'll have to be manually installed into the main libexec directory to work.

**Comment by kaniini** [comment]
@CosmicToast do you have any objection to using this script in 0.12?  It looks like it does basically what I would like to see.

**Comment by CosmicToast** [comment]
@kaniini no problem whatsoever.

In case licensing is a concern, consider the license of the script to be as follows:
Any-of (at the choice of the user, alphabetical order):
* [0BSD](https://spdx.org/licenses/0BSD.html)
* [BlueOak-1.0.0](https://blueoakcouncil.org/license/1.0.0)
* [CC0-1.0](https://creativecommons.org/publicdomain/zero/1.0/)
* [Unlicense](https://unlicense.org/)

If possible, I'd appreciate adding that as comments to the top, but without an author attribution.
